File: src/Services/Agent/Copilot/CopilotRunContext.php
<?php

/**
 * CopilotRunContext
 *
 * Signed, short-lived authority context minted by PHP and verified by the
 * Python agent sidecar. Carries scoped claims (patient/encounter scope,
 * allowed tools, rate caps, expiry) but never raw PHI such as patient
 * names, DOB, addresses, phone numbers, or free-text chart content.
 *
 * Wire format: a JWT-like compact string of the form
 *
 *     <base64url(canonical_payload_json)>.<base64url(hmac_sha256_signature)>
 *
 * The payload is canonical JSON (RFC 8785-style minimal -- sorted object
 * keys, no whitespace, no trailing newlines) so that PHP and Python
 * produce byte-identical inputs to HMAC-SHA256 and the resulting
 * signatures compare with constant-time equality.
 *
 * Includes a `key_version` claim so secrets can be rotated. Verifiers
 * resolve the secret to use by `key_version`.
 *
 * @package   OpenEMR
 * @link      https://www.open-emr.org
 * @author    OpenEMR Clinical Co-Pilot Sidecar Migration
 * @copyright Copyright (c) 2026 OpenEMR contributors
 * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
 */

declare(strict_types=1);

namespace OpenEMR\Services\Agent\Copilot;

use DomainException;
use JsonException;
use SensitiveParameter;

/**
 * Immutable claim set carried in a signed CopilotRunContext token.
 *
 * The constructor enforces shape invariants. Use {@see self::mint()} to
 * produce a wire string and {@see self::decode()} to inspect an unsigned
 * payload (verification happens in the Python sidecar).
 */
final readonly class CopilotRunContext
{
    /**
     * @param list<string> $allowedTools        Names of tools the runtime may invoke.
     * @param list<string> $allowedSourceTypes  Source-type filters the runtime may query.
     */
    public function __construct(
        public int $userId,
        public string $username,
        public int $patientId,
        public ?int $encounterId,
        public array $allowedTools,
        public array $allowedSourceTypes,
        public int $maxRows,
        public int $lookbackDays,
        public int $expiresAt,
        public string $requestId,
        public string $traceId,
        public string $keyVersion,
    ) {
        if ($this->userId <= 0) {
            throw new DomainException('CopilotRunContext: user_id must be positive');
        }

        if (trim($this->username) === '') {
            throw new DomainException('CopilotRunContext: username is required');
        }

        if ($this->patientId <= 0) {
            throw new DomainException('CopilotRunContext: patient_id must be positive');
        }

        if ($this->encounterId !== null && $this->encounterId <= 0) {
            throw new DomainException('CopilotRunContext: encounter_id must be positive when set');
        }

        foreach ($this->allowedTools as $toolName) {
            if ($toolName === '') {
                throw new DomainException('CopilotRunContext: allowed_tools must be a list of non-empty strings');
            }
        }

        foreach ($this->allowedSourceTypes as $sourceType) {
            if ($sourceType === '') {
                throw new DomainException(
                    'CopilotRunContext: allowed_source_types must be a list of non-empty strings'
                );
            }
        }

        if ($this->maxRows <= 0) {
            throw new DomainException('CopilotRunContext: max_rows must be positive');
        }

        if ($this->lookbackDays <= 0) {
            throw new DomainException('CopilotRunContext: lookback_days must be positive');
        }

        if ($this->expiresAt <= 0) {
            throw new DomainException('CopilotRunContext: expires_at must be a positive unix timestamp');
        }

        if (trim($this->requestId) === '') {
            throw new DomainException('CopilotRunContext: request_id is required');
        }

        if (trim($this->traceId) === '') {
            throw new DomainException('CopilotRunContext: trace_id is required');
        }

        if (trim($this->keyVersion) === '') {
            throw new DomainException('CopilotRunContext: key_version is required');
        }
    }

    /**
     * Mint a signed wire string from the supplied claim array.
     *
     * Required keys in `$claims`:
     *   user_id, username, patient_id, encounter_id (nullable),
     *   allowed_tools, allowed_source_types, max_rows, lookback_days,
     *   expires_at, request_id, trace_id.
     *
     * `key_version` is supplied separately so callers explicitly choose the
     * key generation rather than letting input data drive it.
     *
     * @param array<string, mixed> $claims Authority claims to embed.
     */
    public static function mint(
        array $claims,
        #[SensitiveParameter] string $secret,
        string $keyVersion,
    ): string {
        if ($secret === '') {
            throw new DomainException('CopilotRunContext: secret must not be empty');
        }

        if (trim($keyVersion) === '') {
            throw new DomainException('CopilotRunContext: key_version must not be empty');
        }

        // Build a CopilotRunContext to enforce shape invariants up front. We
        // re-extract the canonical payload from the value object so that
        // decode/round-trip semantics match exactly.
        $context = self::fromClaimsArray($claims, $keyVersion);
        $payloadJson = self::canonicalPayloadJson($context->toClaimsArray());
        $signature = hash_hmac('sha256', $payloadJson, $secret, true);

        return self::base64UrlEncode($payloadJson) . '.' . self::base64UrlEncode($signature);
    }

    /**
     * Decode a wire string into its claims array WITHOUT verifying the signature.
     *
     * This is for inspection only -- never trust these claims for
     * authorization decisions. The Python sidecar performs the
     * authoritative HMAC verification.
     *
     * @return array<string, mixed>
     *
     * @throws DomainException When the wire format is malformed.
     */
    public static function decode(string $wire): array
    {
        $parts = explode('.', $wire);
        if (count($parts) !== 2) {
            throw new DomainException('CopilotRunContext: wire format must be <payload>.<signature>');
        }

        [$encodedPayload, $encodedSignature] = $parts;
        if ($encodedPayload === '' || $encodedSignature === '') {
            throw new DomainException('CopilotRunContext: wire payload or signature segment is empty');
        }

        $payloadJson = self::base64UrlDecode($encodedPayload);
        if ($payloadJson === null) {
            throw new DomainException('CopilotRunContext: payload segment is not valid base64url');
        }

        try {
            /** @var mixed $decoded */
            $decoded = json_decode($payloadJson, true, 32, JSON_THROW_ON_ERROR);
        } catch (JsonException) {
            throw new DomainException('CopilotRunContext: payload is not valid JSON');
        }

        if (!is_array($decoded)) {
            throw new DomainException('CopilotRunContext: payload must decode to an object');
        }

        /** @var array<string, mixed> $decoded */
        return $decoded;
    }

    /**
     * @return array<string, mixed>
     */
    public function toClaimsArray(): array
    {
        return [
            'allowed_source_types' => $this->allowedSourceTypes,
            'allowed_tools' => $this->allowedTools,
            'encounter_id' => $this->encounterId,
            'expires_at' => $this->expiresAt,
            'key_version' => $this->keyVersion,
            'lookback_days' => $this->lookbackDays,
            'max_rows' => $this->maxRows,
            'patient_id' => $this->patientId,
            'request_id' => $this->requestId,
            'trace_id' => $this->traceId,
            'user_id' => $this->userId,
            'username' => $this->username,
        ];
    }

    /**
     * Build a canonical JSON encoding of the supplied claims array.
     *
     * "Canonical" here means: object keys sorted lexicographically at every
     * level, no insignificant whitespace, slashes left unescaped, unicode
     * left unescaped. PHP's JSON_UNESCAPED_SLASHES + JSON_UNESCAPED_UNICODE
     * matches Python's `json.dumps(..., separators=(",", ":"), sort_keys=True,
     * ensure_ascii=False)`.
     *
     * @param array<string, mixed> $claims
     */
    private static function canonicalPayloadJson(array $claims): string
    {
        $sorted = self::deepSortByKey($claims);

        try {
            return json_encode(
                $sorted,
                JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE,
            );
        } catch (JsonException $e) {
            throw new DomainException('CopilotRunContext: failed to encode canonical JSON', 0, $e);
        }
    }

    /**
     * Recursively sort associative-array keys so canonical JSON is deterministic.
     *
     * Lists (numeric-keyed arrays) preserve element order; only object-like
     * keys are sorted.
     */
    private static function deepSortByKey(mixed $value): mixed
    {
        if (!is_array($value)) {
            return $value;
        }

        $isList = array_is_list($value);
        $result = [];

        if ($isList) {
            foreach ($value as $item) {
                $result[] = self::deepSortByKey($item);
            }

            return $result;
        }

        $keys = array_keys($value);
        sort($keys, SORT_STRING);
        foreach ($keys as $key) {
            $result[$key] = self::deepSortByKey($value[$key]);
        }

        return $result;
    }

    /**
     * @param array<string, mixed> $claims
     */
    private static function fromClaimsArray(array $claims, string $keyVersion): self
    {
        return new self(
            userId: self::requireInt($claims, 'user_id'),
            username: self::requireString($claims, 'username'),
            patientId: self::requireInt($claims, 'patient_id'),
            encounterId: self::optionalInt($claims, 'encounter_id'),
            allowedTools: self::requireListOfStrings($claims, 'allowed_tools'),
            allowedSourceTypes: self::requireListOfStrings($claims, 'allowed_source_types'),
            maxRows: self::requireInt($claims, 'max_rows'),
            lookbackDays: self::requireInt($claims, 'lookback_days'),
            expiresAt: self::requireInt($claims, 'expires_at'),
            requestId: self::requireString($claims, 'request_id'),
            traceId: self::requireString($claims, 'trace_id'),
            keyVersion: $keyVersion,
        );
    }

    /**
     * @param array<string, mixed> $claims
     */
    private static function requireInt(array $claims, string $key): int
    {
        if (!array_key_exists($key, $claims)) {
            throw new DomainException("CopilotRunContext: missing required claim '{$key}'");
        }

        $value = $claims[$key];
        if (!is_int($value)) {
            throw new DomainException("CopilotRunContext: claim '{$key}' must be an int");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $claims
     */
    private static function optionalInt(array $claims, string $key): ?int
    {
        if (!array_key_exists($key, $claims)) {
            return null;
        }

        $value = $claims[$key];
        if ($value === null) {
            return null;
        }

        if (!is_int($value)) {
            throw new DomainException("CopilotRunContext: claim '{$key}' must be an int or null");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $claims
     */
    private static function requireString(array $claims, string $key): string
    {
        if (!array_key_exists($key, $claims)) {
            throw new DomainException("CopilotRunContext: missing required claim '{$key}'");
        }

        $value = $claims[$key];
        if (!is_string($value)) {
            throw new DomainException("CopilotRunContext: claim '{$key}' must be a string");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $claims
     *
     * @return list<string>
     */
    private static function requireListOfStrings(array $claims, string $key): array
    {
        if (!array_key_exists($key, $claims)) {
            throw new DomainException("CopilotRunContext: missing required claim '{$key}'");
        }

        $value = $claims[$key];
        if (!is_array($value) || !array_is_list($value)) {
            throw new DomainException("CopilotRunContext: claim '{$key}' must be a list");
        }

        $result = [];
        foreach ($value as $item) {
            if (!is_string($item)) {
                throw new DomainException("CopilotRunContext: claim '{$key}' must contain only strings");
            }
            $result[] = $item;
        }

        return $result;
    }

    private static function base64UrlEncode(string $bytes): string
    {
        return rtrim(strtr(base64_encode($bytes), '+/', '-_'), '=');
    }

    private static function base64UrlDecode(string $encoded): ?string
    {
        $remainder = strlen($encoded) % 4;
        if ($remainder !== 0) {
            $encoded .= str_repeat('=', 4 - $remainder);
        }

        $decoded = base64_decode(strtr($encoded, '-_', '+/'), true);
        if ($decoded === false) {
            return null;
        }

        return $decoded;
    }
}

File: agent-service/agent_service/auth/copilot_run_context.py
"""Verifier for the signed CopilotRunContext minted by the PHP host (M3).

The PHP minter (``OpenEMR\\Services\\Agent\\Copilot\\CopilotRunContext``)
emits a JWT-like compact wire string::

    <base64url(canonical_payload_json)>.<base64url(hmac_sha256_signature)>

The payload is canonical JSON: object keys sorted lexicographically at every
level, no insignificant whitespace, ``ensure_ascii=False`` (i.e. unescaped
unicode), unescaped slashes. PHP and Python therefore produce byte-identical
HMAC inputs.

Verification semantics:

* Resolve the secret to use via the ``key_version`` claim (callers supply a
  ``secret_resolver``). Unknown versions are rejected -- the verifier never
  silently falls back.
* Recompute HMAC-SHA256 and compare with :func:`hmac.compare_digest` to
  prevent timing oracles.
* Reject expired tokens (``expires_at < now()``). The clock is injected so
  tests are deterministic.
* Reject malformed wire strings, missing/empty signature segments, missing
  required claims, and tampered payloads.

The verifier returns a Pydantic v2 :class:`CopilotRunContext` on success;
all error paths raise :class:`CopilotRunContextError` with a typed
``reason`` so callers can map failures to fail-closed responses.
"""

from __future__ import annotations

import base64
import binascii
import hashlib
import hmac
import json
from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import StrEnum
from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class CopilotRunContextErrorReason(StrEnum):
    """Discriminator for verification failures.

    The values map to the reasons enumerated by the M3 spec: ``expired``,
    ``tampered``, ``bad_signature``, ``unknown_key_version``, ``malformed``.
    """

    EXPIRED = "expired"
    TAMPERED = "tampered"
    BAD_SIGNATURE = "bad_signature"
    UNKNOWN_KEY_VERSION = "unknown_key_version"
    MALFORMED = "malformed"


class CopilotRunContextError(Exception):
    """Raised when a CopilotRunContext token fails verification.

    Carries a machine-readable :class:`CopilotRunContextErrorReason` so the
    sidecar can return a fail-closed response without leaking which check
    failed at the message-string level.
    """

    def __init__(self, reason: CopilotRunContextErrorReason, message: str) -> None:
        super().__init__(message)
        self.reason: CopilotRunContextErrorReason = reason


# ---------------------------------------------------------------------------
# Validated claim model
# ---------------------------------------------------------------------------


class CopilotRunContext(BaseModel):
    """Validated authority claims unpacked from a verified token.

    Every field corresponds to a claim minted by the PHP host. The model is
    strict: extra keys are rejected and types are not coerced. The Pydantic
    validators enforce shape invariants that mirror the PHP value object.
    """

    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)

    user_id: Annotated[int, Field(gt=0, description="OpenEMR user ID")]
    username: Annotated[str, Field(min_length=1, description="OpenEMR username")]
    patient_id: Annotated[int, Field(gt=0, description="OpenEMR patient PID")]
    encounter_id: int | None = Field(default=None, description="OpenEMR encounter ID, or null")
    allowed_tools: list[str] = Field(description="Names of tools this run may invoke")
    allowed_source_types: list[str] = Field(description="Source-type filters allowed for this run")
    max_rows: Annotated[int, Field(gt=0, description="Per-tool row cap")]
    lookback_days: Annotated[int, Field(gt=0, description="Per-tool history window")]
    expires_at: Annotated[int, Field(gt=0, description="Unix timestamp when this token expires")]
    request_id: Annotated[str, Field(min_length=1, description="Per-request UUID")]
    trace_id: Annotated[str, Field(min_length=1, description="Distributed trace correlation ID")]
    key_version: Annotated[str, Field(min_length=1, description="Secret key generation, e.g. 'v1'")]

    @field_validator("encounter_id")
    @classmethod
    def _encounter_id_must_be_positive_when_set(cls, value: int | None) -> int | None:
        if value is not None and value <= 0:
            msg = "encounter_id must be positive when set"
            raise ValueError(msg)
        return value

    @field_validator("allowed_tools", "allowed_source_types")
    @classmethod
    def _strings_must_be_non_empty(cls, value: list[str]) -> list[str]:
        for item in value:
            if not isinstance(item, str) or item == "":
                msg = "list entries must be non-empty strings"
                raise ValueError(msg)
        return value


# ---------------------------------------------------------------------------
# Verifier
# ---------------------------------------------------------------------------


SecretResolver = Callable[[str], str | None]
"""Maps a ``key_version`` to its shared secret, or ``None`` when unknown.

Returning ``None`` is the canonical way to reject an unknown key version.
The verifier raises :class:`CopilotRunContextError` with reason
``unknown_key_version`` when the resolver returns ``None`` or empty string.
"""


@dataclass(frozen=True, slots=True)
class _DecodedWire:
    """Internal split of a wire token into its three structural pieces."""

    payload_bytes: bytes
    signature_bytes: bytes
    claims: dict[str, Any]


def _utc_now_seconds() -> int:
    """Return the current UTC time as an int unix timestamp.

    Defined at module scope so tests can monkeypatch a deterministic clock
    without instantiating any fixtures.
    """
    return int(datetime.now(tz=timezone.utc).timestamp())


def _b64url_decode(segment: str) -> bytes:
    """Decode an unpadded base64url segment, padding internally."""
    padding = (-len(segment)) % 4
    try:
        return base64.urlsafe_b64decode(segment + ("=" * padding))
    except (binascii.Error, ValueError) as exc:
        msg = "wire segment is not valid base64url"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg) from exc


def _canonical_payload(claims: dict[str, Any]) -> bytes:
    """Re-encode claims to canonical JSON for signature verification.

    Matches the PHP minter: sorted keys, no whitespace, unicode unescaped.
    Lists preserve order; only dicts have keys sorted (``sort_keys=True``
    handles every nesting level).
    """
    return json.dumps(
        claims,
        separators=(",", ":"),
        sort_keys=True,
        ensure_ascii=False,
    ).encode("utf-8")


def _split_wire(wire: str) -> _DecodedWire:
    """Split a wire string into payload + signature, decoding both."""
    if not isinstance(wire, str) or wire == "":
        msg = "wire token must be a non-empty string"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    parts = wire.split(".")
    if len(parts) != 2:
        msg = "wire format must be <payload>.<signature>"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    payload_segment, signature_segment = parts
    if payload_segment == "" or signature_segment == "":
        msg = "wire payload or signature segment is empty"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    payload_bytes = _b64url_decode(payload_segment)
    signature_bytes = _b64url_decode(signature_segment)

    try:
        decoded = json.loads(payload_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        msg = "payload is not valid JSON"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg) from exc

    if not isinstance(decoded, dict):
        msg = "payload must decode to a JSON object"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    return _DecodedWire(
        payload_bytes=payload_bytes,
        signature_bytes=signature_bytes,
        claims=decoded,
    )


def verify_copilot_run_context(
    wire: str,
    secret_resolver: SecretResolver,
    *,
    now: Callable[[], int] = _utc_now_seconds,
) -> CopilotRunContext:
    """Verify and parse a CopilotRunContext wire token.

    Parameters
    ----------
    wire:
        The compact wire string produced by the PHP minter.
    secret_resolver:
        Callable invoked with the ``key_version`` claim; must return the
        shared secret bytes-as-str, or ``None`` to signal "unknown version".
    now:
        Injected clock returning the current unix timestamp. Defaults to
        UTC ``time.time()``-equivalent. Tests pass a deterministic stub.

    Returns
    -------
    CopilotRunContext
        The validated, frozen Pydantic model.

    Raises
    ------
    CopilotRunContextError
        With ``reason`` set to one of:
        ``malformed``, ``unknown_key_version``, ``bad_signature``,
        ``tampered``, or ``expired``. Callers should map every reason to a
        fail-closed 401/403 response and never leak the wire claims to the
        client.
    """
    decoded = _split_wire(wire)
    claims = decoded.claims

    key_version = claims.get("key_version")
    if not isinstance(key_version, str) or key_version == "":
        msg = "missing or empty key_version claim"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    secret = secret_resolver(key_version)
    if not secret:
        msg = f"unknown key_version: {key_version!r}"
        raise CopilotRunContextError(CopilotRunContextErrorReason.UNKNOWN_KEY_VERSION, msg)

    expected_signature = hmac.new(
        secret.encode("utf-8"),
        _canonical_payload(claims),
        hashlib.sha256,
    ).digest()

    if not hmac.compare_digest(expected_signature, decoded.signature_bytes):
        # We cannot tell whether the payload was modified or the secret was
        # wrong without distinguishing trust models, so we default to the
        # "tampered" reason. Callers can decide whether to differentiate.
        msg = "signature does not match payload + secret"
        raise CopilotRunContextError(CopilotRunContextErrorReason.TAMPERED, msg)

    expires_at = claims.get("expires_at")
    if not isinstance(expires_at, int) or expires_at <= 0:
        msg = "missing or invalid expires_at claim"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg)

    if expires_at < now():
        msg = "token expired"
        raise CopilotRunContextError(CopilotRunContextErrorReason.EXPIRED, msg)

    try:
        return CopilotRunContext.model_validate(claims)
    except ValueError as exc:
        # Pydantic validation failure on a token that already passed HMAC
        # means the structure was malformed at mint time.
        msg = f"claims failed schema validation: {exc}"
        raise CopilotRunContextError(CopilotRunContextErrorReason.MALFORMED, msg) from exc
